src/Microsoft.AspNet.PipelineCore/FormFeature.cs

```diff
@@ -1,71 +1,200 @@
 // Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
 using System.Collections.Generic;
 using System.IO;
+using System.Linq;
 using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.AspNet.FeatureModel;
 using Microsoft.AspNet.Http;
-using Microsoft.AspNet.HttpFeature;
-using Microsoft.AspNet.PipelineCore.Infrastructure;
+using Microsoft.AspNet.PipelineCore.Collections;
 using Microsoft.AspNet.WebUtilities;
-using Microsoft.AspNet.WebUtilities.Collections;
 
 namespace Microsoft.AspNet.PipelineCore
 {
     public class FormFeature : IFormFeature
     {
-        private readonly IFeatureCollection _features;
-        private readonly FeatureReference<IHttpRequestFeature> _request = FeatureReference<IHttpRequestFeature>.Default;
-        private Stream _bodyStream;
-        private IReadableStringCollection _form;
+        private readonly HttpRequest _request;
 
-        public FormFeature([NotNull] IDictionary<string, string[]> form)
-            : this (new ReadableStringCollection(form))
+        public FormFeature([NotNull] IFormCollection form)
         {
+            Form = form;
         }
 
-        public FormFeature([NotNull] IReadableStringCollection form)
+        public FormFeature([NotNull] HttpRequest request)
         {
-            _form = form;
+            _request = request;
         }
 
-        public FormFeature([NotNull] IFeatureCollection features)
+        public bool HasFormContentType
         {
-            _features = features;
+            get
+            {
+                // Set directly
+                if (Form != null)
+                {
+                    return true;
+                }
+
+                return HasApplicationFormContentType() || HasMultipartFormContentType();
+            }
+        }
+
+        public IFormCollection Form { get; set; }
+
+        public IFormCollection ReadForm()
+        {
+            if (Form != null)
```

Comment:
There's nothing nearly as expensive elsewhere on the request or response yet. However, there are plenty of things on both request and response that are not thread safe. Reading the request body, writing the response body, setting response headers, etc.. Even reading the request headers is not guaranteed to be thread safe.